cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs

```diff
@@ -0,0 +1,624 @@
+runGenesisKeyGenDelegateCmd
+  :: GenesisKeyGenDelegateCmdArgs
+  -> ExceptT GenesisCmdError IO ()
+runGenesisKeyGenDelegateCmd
+    Cmd.GenesisKeyGenDelegateCmdArgs
+    { Cmd.verificationKeyPath
+    , Cmd.signingKeyPath
+    , Cmd.opCertCounterPath
+    } = do
+    skey <- liftIO $ generateSigningKey AsGenesisDelegateKey
+    let vkey = getVerificationKey skey
+    firstExceptT GenesisCmdGenesisFileError
+      . newExceptT
+      $ writeLazyByteStringFile signingKeyPath
+      $ textEnvelopeToJSON (Just skeyDesc) skey
+    firstExceptT GenesisCmdGenesisFileError
+      . newExceptT
+      $ writeLazyByteStringFile verificationKeyPath
+      $ textEnvelopeToJSON (Just Key.genesisVkeyDelegateDesc) vkey
+    firstExceptT GenesisCmdGenesisFileError
+      . newExceptT
+      $ writeLazyByteStringFile opCertCounterPath
+      $ textEnvelopeToJSON (Just certCtrDesc)
+      $ OperationalCertificateIssueCounter
+          initialCounter
+          (castVerificationKey vkey)  -- Cast to a 'StakePoolKey'
+  where
+    skeyDesc, certCtrDesc :: TextEnvelopeDescr
+    skeyDesc = "Genesis delegate operator key"
+    certCtrDesc = "Next certificate issue number: "
+               <> fromString (show initialCounter)
+    initialCounter :: Word64
+    initialCounter = 0
+
+
+runGenesisKeyGenDelegateVRF ::
+     VerificationKeyFile Out
+  -> SigningKeyFile Out
+  -> ExceptT GenesisCmdError IO ()
+runGenesisKeyGenDelegateVRF vkeyPath skeyPath = do
+    skey <- liftIO $ generateSigningKey AsVrfKey
+    let vkey = getVerificationKey skey
+    firstExceptT GenesisCmdGenesisFileError
+      . newExceptT
+      $ writeLazyByteStringFile skeyPath
+      $ textEnvelopeToJSON (Just skeyDesc) skey
+    firstExceptT GenesisCmdGenesisFileError
+      . newExceptT
+      $ writeLazyByteStringFile vkeyPath
+      $ textEnvelopeToJSON (Just vkeyDesc) vkey
+  where
+    skeyDesc, vkeyDesc :: TextEnvelopeDescr
+    skeyDesc = "VRF Signing Key"
+    vkeyDesc = "VRF Verification Key"
+
+
+runGenesisCreateTestNetDataCmd
+  :: GenesisCreateTestNetDataCmdArgs
+  -> ExceptT GenesisCmdError IO ()
+runGenesisCreateTestNetDataCmd Cmd.GenesisCreateTestNetDataCmdArgs
+  {  networkId
+   , specShelley
+   , numGenesisKeys
+   , numPools
+   , numStakeDelegators
+   , numStuffedUtxo
+   , numUtxoKeys
+   , supply
+   , supplyDelegated
+   , systemStart
+   , outputDir }
+   = do
+  liftIO $ createDirectoryIfMissing False outputDir
+  shelleyGenesis <-
+    case specShelley of
+      Just shelleyPath ->
+        newExceptT $ readAndDecodeShelleyGenesis shelleyPath
+      Nothing -> do
+        -- No template given: a default file is created
+        let newTemplate = shelleyGenesisDefaults { sgNetworkMagic = unNetworkMagic (toNetworkMagic networkId) }
+            destination = outputDir </> "genesis-shelley.json"
+        liftIO $ LBS.writeFile destination $ encode newTemplate
+        pure newTemplate
+
+  -- Genesis keys
+  let mkPaths :: String -> String -> String -> Map Int FilePath =
```

Comment:
It would be clearer to factor this out IMO and have a little comment as to what the map will be used for. Could be done in a follow up PR.

Reply:
Done :+1: